File: go-backend-interview-prep/07_io_files_networking/05_env_and_args/main.go
package main

import (
	"flag"
	"fmt"
	"os"
	"strings"
)

func main() {
	// --- Example 1: os.Args ---
	fmt.Println("=== os.Args ===")
	fmt.Println("  program:", os.Args[0])
	fmt.Println("  all args:", os.Args[1:])

	// --- Example 2: flag package ---
	fmt.Println("\n=== flag package ===")
	name := flag.String("name", "world", "who to greet")
	count := flag.Int("count", 1, "how many times")
	verbose := flag.Bool("v", false, "verbose output")
	flag.Parse()

	fmt.Printf("  name=%q count=%d verbose=%v\n", *name, *count, *verbose)
	fmt.Println("  remaining args:", flag.Args())

	for i := 0; i < *count; i++ {
		fmt.Printf("  hello, %s!\n", *name)
	}

	// --- Example 3: environment variables ---
	fmt.Println("\n=== Environment variables ===")

	// Getenv returns "" if unset
	home := os.Getenv("HOME")
	fmt.Println("  HOME:", home)

	// LookupEnv distinguishes unset from empty
	val, exists := os.LookupEnv("MY_APP_SECRET")
	if exists {
		fmt.Println("  MY_APP_SECRET:", val)
	} else {
		fmt.Println("  MY_APP_SECRET: (not set)")
	}

	// List env vars matching a prefix
	fmt.Println("\n=== Env vars starting with 'GO' ===")
	for _, env := range os.Environ() {
		if strings.HasPrefix(env, "GO") {
			fmt.Println(" ", env)
		}
	}
}

File: go-backend-interview-prep/07_io_files_networking/05_env_and_args/main.py
"""Environment variables and command-line arguments -- Python equivalent."""

import argparse
import os
import sys


def main() -> None:
    # --- Example 1: sys.argv ---
    print("=== sys.argv ===")
    print(f"  program: {sys.argv[0]}")
    print(f"  all args: {sys.argv[1:]}")

    # --- Example 2: argparse ---
    print("\n=== argparse ===")
    parser = argparse.ArgumentParser(description="Greeting demo")
    parser.add_argument("--name", default="world", help="who to greet")
    parser.add_argument("--count", type=int, default=1, help="how many times")
    parser.add_argument("-v", action="store_true", help="verbose output")
    args, remaining = parser.parse_known_args()

    print(f"  name={args.name!r} count={args.count} verbose={args.v}")
    print(f"  remaining args: {remaining}")

    for _ in range(args.count):
        print(f"  hello, {args.name}!")

    # --- Example 3: environment variables ---
    print("\n=== Environment variables ===")

    # os.environ.get returns default if unset
    home = os.environ.get("HOME", "")
    print(f"  HOME: {home}")

    # Check existence explicitly
    secret = os.environ.get("MY_APP_SECRET")
    if secret is not None:
        print(f"  MY_APP_SECRET: {secret}")
    else:
        print("  MY_APP_SECRET: (not set)")

    # List env vars matching a prefix
    print("\n=== Env vars starting with 'GO' ===")
    for key, val in sorted(os.environ.items()):
        if key.startswith("GO"):
            print(f"  {key}={val}")


if __name__ == "__main__":
    main()
